### PR TITLE
Add default trait implementations for "c-unwind" ABI function pointers

### PR DESCRIPTION
**Pull request by lopopolo**
Following up on #92964, only add default trait implementations for the `c-unwind` family of function pointers. The previous attempt in #92964 added trait implementations for many more ABIs and ran into concerns regarding the increase in size of the libcore rlib.

An attempt to abstract away function pointer types behind a unified trait to reduce the duplication of trait impls is being discussed in #99531 but this change looks to be blocked on a lang MCP.

Following @RalfJung's suggestion in https://github.com/rust-lang/rust/pull/99531#issuecomment-1233440142, this commit is another cut at #92964 but it _only_ adds the impls for `extern "C-unwind" fn` and `unsafe extern "C-unwind" fn`.

I am interested in landing this patch to unblock the stabilization of the `c_unwind` feature.

RFC: https://github.com/rust-lang/rfcs/pull/2945
Tracking Issue: https://github.com/rust-lang/rust/issues/74990

### CODE REVIEW

**Comment by lopopolo** [comment]
it looks like I need to bless changes to the UI tests. Let me try and figure out how to do that 😄

**Comment by RalfJung** [comment]
Yeah, as expected that extra metadata in the rlib makes everything measurably slower. Not sure if there is a way that we can avoid some of these instances (do we really need all the traits here), but in the end it boils down to judging whether that's a cost we are willing to accept for the benefit of https://github.com/rust-lang/rust/issues/74990.

Sounds like a @rust-lang/compiler question to me.

**Comment by pnkfelix** [comment]
Discussed in T-compiler [triage meeting](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202022-09-08/near/297801084).

I think we should move forward with this, despite the performance hit. I don't want to block c-unwind based on the performance regression observed above.

(If it turns out that the performance impact matters, then we can adopt other approaches, such as the one suggested in #99531; though I'm hoping that will not be necessary solely for this issue.)

@rustbot label: -I-compiler-nominated

**Comment by pnkfelix** [comment]
(also, as a side note: the primary benchmarks only regressed on "doc full" scenarios. Normal compilation (check/debug/opt) all came out looking great I would say.)

@rustbot label: +perf-regression-triaged

**Comment by lopopolo** [comment]
@RalfJung Given the above triage on the perf regression from the compiler team, what are the next steps for getting this approved and merged?

**Comment by RalfJung** [comment]
It needs someone from t-libs to review and approve.
r? libs

**Review by thomcc — APPROVED** [review]
Impl looks fine, I think this needs libs-api signoff, or perhaps an ACP.

**Comment by thomcc** [comment]
Unless that's already happened somewhere?

**Comment by lopopolo** [comment]
@thomcc are these next steps something you need me to drive or is it something you can help me with?

r? T-libs-api

**Comment by lopopolo** [comment]
@thomcc are these next steps something you need me to drive or is it something you can help me with?

r? libs-api

**Comment by lopopolo** [comment]
I'm not sure how to get rustbot to request a review from libs-api but if that's what needs to happen, @thomcc can you assign someone?

**Comment by thomcc** [comment]
If this hasn't been signed off by libs-api you need to do an ACP, I think. https://std-dev-guide.rust-lang.org/feature-lifecycle/api-change-proposals.html

**Comment by lopopolo** [comment]
ACP is here:

- https://github.com/rust-lang/libs-team/issues/100

**Comment by lopopolo** [comment]
Hi folks, checking in here. Is there anything I can do to unblock this PR?

**Comment by RalfJung** [comment]
@rust-lang/libs-api I assume the next step is for [the ACP](https://github.com/rust-lang/libs-team/issues/100) to be discussed? Not sure how that process looks like.

**Comment by lopopolo** [comment]
Hi folks, there hasn't been any movement either in this PR or in the associated ACP. I am not sure what to do next or whose approval I am waiting on.

**Comment by lopopolo** [comment]
@thomcc as the assignee of this ticket, can you suggest some next steps or help me get the right eyes on this ticket and the ACP so this PR can be merged?

**Comment by RalfJung** [comment]
@rust-lang/libs-api see the PR description for what the PR is doing and why. T-compiler approved the perf regression. The PR author was asked to file an ACP, which [they did](https://github.com/rust-lang/libs-team/issues/100), and since then the PR has been stalling. Looks like the team needs to make a decision so that we can make progress here. :)

**Comment by thomcc** [comment]
We discussed this in the most recent T-libs-api meeting and were fine with it. I think it needs FCP though. Can someone start that?

**Comment by lopopolo** [comment]
Is the FCP more appropriate for the stabilization PR? All of these function pointer types are unstable.

**Comment by thomcc** [comment]
Oh. I guess that's true. Since it's still marked as nominated I'll bring it up again in the meeting to ask if we want to FCP a case like this.

**Comment by m-ou-se** [comment]
@thomcc Feel free to r+ without an FCP, since the types are still unstable.

The macro expands to impls with a `#[stable]` tag though. It'd be nice to fix that, but maybe not worth the effort.

**Comment by thomcc** [comment]
> It'd be nice to fix that, but maybe not worth the effort.

We'd probably want it to have an accurate stability tag when we finally do stabilize it, rather than one that says 1.4.0. So we will need to adjust how the attribute is handled no matter what.

So I think I'd like this to fix that to say unstable for now, even though its a bit of a pain, then I'll r+.

@rustbot author

**Comment by lopopolo** [comment]
@thomcc I attempted to try making these changes and got these errors due to https://github.com/rust-lang/rust/issues/55436:

```
error: an `#[unstable]` annotation here has no effect
    --> library/core/src/ptr/mod.rs:1846:9
     |
1832 | / macro_rules! maybe_fnptr_doc {
1833 | |     (@ #[$meta:meta] $item:item) => {
1834 | |         #[doc(hidden)]
1835 | |         #[$meta]
...    |
1846 | |         #[$meta]
     | |         ^^^^^^^^
1847 | |         $item
1848 | |     };
1849 | | }
     | |_- in this expansion of `maybe_fnptr_doc!` (#3)
...
1857 | / macro_rules! fnptr_impls_safety_abi {
1858 | |     (#[$meta:meta] $FnTy: ty, $($Arg: ident),*) => {
1859 | |         maybe_fnptr_doc! {
1860 | |             $($Arg)* @
...    |
1918 | /         maybe_fnptr_doc! {
1919 |               $($Arg)* @
1920 |               #[$meta]
1921 |               impl<Ret, $($Arg),*> fmt::Debug for $FnTy {
...
1925 |               }
1926 | |         }
     | |_________- in this macro invocation (#3)
1927 | |     }
1928 | | }
     | |_- in this expansion of `fnptr_impls_safety_abi!` (#2)
1929 |
1930 | / macro_rules! fnptr_impls_args {
1931 | |     ($($Arg: ident),+) => {
1932 | |         fnptr_impls_safety_abi! { #[stable(feature = "fnptr_impls", since = "1.4.0")] extern "Rust" fn($($Arg),+) -> Ret, $($Arg),+ }
1933 | |         fnptr_impls_safety_abi! { #[stable(feature = "fnptr_impls", since = "1.4.0")] extern "C" fn($($Arg),+) -> Ret, $($Arg),+ }
...    |
1936 | |         fnptr_impls_safety_abi! { #[unstable(feature = "c_unwind", issue = "74990")] extern "C-unwind" fn($($Arg),+ , ...) -> Ret, $($Arg),+ }
     | |         -------------------------------------------------------------------------------------------------------------------------------------- in this macro invocation (#2)
...    |
1951 | |     };
1952 | | }
     | |_- in this expansion of `fnptr_impls_args!` (#1)
...
1966 |   fnptr_impls_args! { A, B, C, D, E, F, G, H, I, J, K, L }
     |   -------------------------------------------------------- in this macro invocation (#1)
     |
     = note: see issue #55436 <https://github.com/rust-lang/rust/issues/55436> for more information
```

My suggestion (and desire so this PR can be merged) would be to rework the stability annotations as part of the stabilization PR.

## Patch

```diff
diff --git i/library/core/src/ptr/mod.rs w/library/core/src/ptr/mod.rs
index 8d3f4e8f569..74a91394806 100644
--- i/library/core/src/ptr/mod.rs
+++ w/library/core/src/ptr/mod.rs
@@ -1855,10 +1855,10 @@ macro_rules! maybe_fnptr_doc {
 
 // Impls for function pointers
 macro_rules! fnptr_impls_safety_abi {
-    ($FnTy: ty, $($Arg: ident),*) => {
+    (#[$meta:meta] $FnTy: ty, $($Arg: ident),*) => {
         maybe_fnptr_doc! {
             $($Arg)* @
-            #[stable(feature = "fnptr_impls", since = "1.4.0")]
+            #[$meta]
             impl<Ret, $($Arg),*> PartialEq for $FnTy {
                 #[inline]
                 fn eq(&self, other: &Self) -> bool {
@@ -1869,13 +1869,13 @@ fn eq(&self, other: &Self) -> bool {
 
         maybe_fnptr_doc! {
             $($Arg)* @
-            #[stable(feature = "fnptr_impls", since = "1.4.0")]
+            #[$meta]
             impl<Ret, $($Arg),*> Eq for $FnTy {}
         }
 
         maybe_fnptr_doc! {
             $($Arg)* @
-            #[stable(feature = "fnptr_impls", since = "1.4.0")]
+            #[$meta]
             impl<Ret, $($Arg),*> PartialOrd for $FnTy {
                 #[inline]
                 fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
@@ -1886,7 +1886,7 @@ fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
 
         maybe_fnptr_doc! {
             $($Arg)* @
-            #[stable(feature = "fnptr_impls", since = "1.4.0")]
+            #[$meta]
             impl<Ret, $($Arg),*> Ord for $FnTy {
                 #[inline]
                 fn cmp(&self, other: &Self) -> Ordering {
@@ -1897,7 +1897,7 @@ fn cmp(&self, other: &Self) -> Ordering {
 
         maybe_fnptr_doc! {
             $($Arg)* @
-            #[stable(feature = "fnptr_impls", since = "1.4.0")]
+            #[$meta]
             impl<Ret, $($Arg),*> hash::Hash for $FnTy {
                 fn hash<HH: hash::Hasher>(&self, state: &mut HH) {
                     state.write_usize(*self as usize)
@@ -1907,7 +1907,7 @@ fn hash<HH: hash::Hasher>(&self, state: &mut HH) {
 
         maybe_fnptr_doc! {
             $($Arg)* @
-            #[stable(feature = "fnptr_impls", since = "1.4.0")]
+            #[$meta]
             impl<Ret, $($Arg),*> fmt::Pointer for $FnTy {
                 fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
                     fmt::pointer_fmt_inner(*self as usize, f)
@@ -1917,7 +1917,7 @@ fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
 
         maybe_fnptr_doc! {
             $($Arg)* @
-            #[stable(feature = "fnptr_impls", since = "1.4.0")]
+            #[$meta]
             impl<Ret, $($Arg),*> fmt::Debug for $FnTy {
                 fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
                     fmt::pointer_fmt_inner(*self as usize, f)
@@ -1929,25 +1929,25 @@ fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
 
 macro_rules! fnptr_impls_args {
     ($($Arg: ident),+) => {
-        fnptr_impls_safety_abi! { extern "Rust" fn($($Arg),+) -> Ret, $($Arg),+ }
-        fnptr_impls_safety_abi! { extern "C" fn($($Arg),+) -> Ret, $($Arg),+ }
-        fnptr_impls_safety_abi! { extern "C" fn($($Arg),+ , ...) -> Ret, $($Arg),+ }
-        fnptr_impls_safety_abi! { extern "C-unwind" fn($($Arg),+) -> Ret, $($Arg),+ }
-        fnptr_impls_safety_abi! { extern "C-unwind" fn($($Arg),+ , ...) -> Ret, $($Arg),+ }
-        fnptr_impls_safety_abi! { unsafe extern "Rust" fn($($Arg),+) -> Ret, $($Arg),+ }
-        fnptr_impls_safety_abi! { unsafe extern "C" fn($($Arg),+) -> Ret, $($Arg),+ }
-        fnptr_impls_safety_abi! { unsafe extern "C" fn($($Arg),+ , ...) -> Ret, $($Arg),+ }
-        fnptr_impls_safety_abi! { unsafe extern "C-unwind" fn($($Arg),+) -> Ret, $($Arg),+ }
-        fnptr_impls_safety_abi! { unsafe extern "C-unwind" fn($($Arg),+ , ...) -> Ret, $($Arg),+ }
+        fnptr_impls_safety_abi! { #[stable(feature = "fnptr_impls", since = "1.4.0")] extern "Rust" fn($($Arg),+) -> Ret, $($Arg),+ }
+        fnptr_impls_safety_abi! { #[stable(feature = "fnptr_impls", since = "1.4.0")] extern "C" fn($($Arg),+) -> Ret, $($Arg),+ }
+        fnptr_impls_safety_abi! { #[stable(feature = "fnptr_impls", since = "1.4.0")] extern "C" fn($($Arg),+ , ...) -> Ret, $($Arg),+ }
+        fnptr_impls_safety_abi! { #[unstable(feature = "c_unwind", issue = "74990")] extern "C-unwind" fn($($Arg),+) -> Ret, $($Arg),+ }
+        fnptr_impls_safety_abi! { #[unstable(feature = "c_unwind", issue = "74990")] extern "C-unwind" fn($($Arg),+ , ...) -> Ret, $($Arg),+ }
+        fnptr_impls_safety_abi! { #[stable(feature = "fnptr_impls", since = "1.4.0")] unsafe extern "Rust" fn($($Arg),+) -> Ret, $($Arg),+ }
+        fnptr_impls_safety_abi! { #[stable(feature = "fnptr_impls", since = "1.4.0")] unsafe extern "C" fn($($Arg),+) -> Ret, $($Arg),+ }
+        fnptr_impls_safety_abi! { #[stable(feature = "fnptr_impls", since = "1.4.0")] unsafe extern "C" fn($($Arg),+ , ...) -> Ret, $($Arg),+ }
+        fnptr_impls_safety_abi! { #[stable(feature = "fnptr_impls", since = "1.4.0")] unsafe extern "C-unwind" fn($($Arg),+) -> Ret, $($Arg),+ }
+        fnptr_impls_safety_abi! { #[stable(feature = "fnptr_impls", since = "1.4.0")] unsafe extern "C-unwind" fn($($Arg),+ , ...) -> Ret, $($Arg),+ }
     };
     () => {
         // No variadic functions with 0 parameters
-        fnptr_impls_safety_abi! { extern "Rust" fn() -> Ret, }
-        fnptr_impls_safety_abi! { extern "C" fn() -> Ret, }
-        fnptr_impls_safety_abi! { extern "C-unwind" fn() -> Ret, }
-        fnptr_impls_safety_abi! { unsafe extern "Rust" fn() -> Ret, }
-        fnptr_impls_safety_abi! { unsafe extern "C" fn() -> Ret, }
-        fnptr_impls_safety_abi! { unsafe extern "C-unwind" fn() -> Ret, }
+        fnptr_impls_safety_abi! { #[stable(feature = "fnptr_impls", since = "1.4.0")] extern "Rust" fn() -> Ret, }
+        fnptr_impls_safety_abi! { #[stable(feature = "fnptr_impls", since = "1.4.0")] extern "C" fn() -> Ret, }
+        fnptr_impls_safety_abi! { #[unstable(feature = "c_unwind", issue = "74990")] extern "C-unwind" fn() -> Ret, }
+        fnptr_impls_safety_abi! { #[stable(feature = "fnptr_impls", since = "1.4.0")] unsafe extern "Rust" fn() -> Ret, }
+        fnptr_impls_safety_abi! { #[stable(feature = "fnptr_impls", since = "1.4.0")] unsafe extern "C" fn() -> Ret, }
+        fnptr_impls_safety_abi! { #[unstable(feature = "c_unwind", issue = "74990")] unsafe extern "C-unwind" fn() -> Ret, }
     };
 }
 ```

**Comment by m-ou-se** [comment]
> error: an `#[unstable]` annotation here has no effect

Ah, yeah. That check isn't very accurate, unfortunately. (Which I'm to blame for, I think.) It probably doesn't understand that `extern "C-unwind" fn()` is unstable, since it isn't a type that's defined anywhere with an #[unstable] tag. (It gives the error when neither the trait nor the type is #[unstable].)

**Comment by m-ou-se** [comment]
(Fixing that now..)

**Comment by m-ou-se** [comment]
https://github.com/rust-lang/rust/pull/103239 fixes that error.

**Comment by lopopolo** [comment]
thanks @m-ou-se. I'll rework my macro changes once your PR makes it to master.

@rustbot author

**Comment by lopopolo** [comment]
I couldn't figure out the right combination of `--stage` and `--keep-stage` flags to pass to `x.py` to check this locally, but I did rebase to include https://github.com/rust-lang/rust/pull/103239 and reworked the macro from my proposed patch above. Let's see if CI passes.

**Comment by lopopolo** [comment]
Looks like the build still fails with `an `#[unstable]` annotation here has no effect`. @m-ou-se do I need to wait for a new bootstrap compiler to be released?

**Comment by m-ou-se** [comment]
> Looks like the build still fails with `an #[unstable] annotation here has no effect` error. @m-ou-se do I need to wait for a new bootstrap compiler to be released?

Oh. :( Yeah. I suppose you could apply `#[cfg(not(bootstrap))]` to line 1868. These trait impls aren't used in the compiler, meaning it's fine that they're missing from the first stage std.

**Comment by lopopolo** [comment]
@thomcc I've made the stability annotation changes. This is ready for review.

@rustbot ready

**Comment by thomcc** [comment]
LGTM, thanks.

@bors r+

**Comment by nnethercote** [comment]
It's all doc perf regressions, and was decided earlier to be acceptable.

@rustbot label: +perf-regression-triaged

**Pull request merged**
